kolibri/core/assets/src/views/sync/SelectDeviceModalGroup/useDevices.js

```diff
@@ -129,51 +143,83 @@ function useDevicesWithFilter(apiParams, filterFunction) {
 }
 
 /**
- * Filters devices to those that also have the specified channel
- * @param channelId
- * @return {{
- *  devices: Ref<NetworkLocation[]>, hasFetched: Ref<bool>, isFetching: Ref<bool>,
- *  filterFailed: Ref<bool>, forceFetch: (function(): Promise<void>)
- * }}
+ * Produces a memoized function that returns a Promise resolving with a boolean for filtering
+ * devices, and automatically clears memoized result if it fails, unless it is a 404
+ *
+ * @param {function(NetworkLocation): Promise<boolean>} filterFunction
+ * @return {(function(NetworkLocation): Promise<boolean>)}
  */
-export function useDevicesWithChannel(channelId = '') {
-  // If channelId is not provided, then we are at top-level import workflow and do not
-  // disable any devices unless it is unavailable
-  const filterDevices =
-    channelId === '' ? () => Promise.resolve(true) : channelIsAvailableAtDevice.bind({}, channelId);
+function useAsyncDeviceFilter(filterFunction) {
+  const memoized = useMemoize(filterFunction, { getKey: device => device.id });
 
-  return useDevicesWithFilter({}, filterDevices);
+  return async function deviceFilter(device) {
+    try {
+      return await memoized(device);
+    } catch (e) {
+      // If not 404 clear cache to try again on next poll
+      if (e?.response?.status !== 404) {
```

Comment:
We can't use the [optional chaining operator](https://caniuse.com/mdn-javascript_operators_optional_chaining) until next release after we lose IE11 support. We're deprecating in this release. We've used `_.get(e, 'response.status')` elsewhere in Kolibri to a similar effect in the mean time.

Reply:
Ah I thought this would get babelified. Will update